### PR TITLE
Fewer loop checks in SkipWhile

### PR DESCRIPTION
**Pull request by jamesqo**
Separated from PR #2318. Sorry if I seem to be making a lot of PRs today haha, I just want to finish up with all of the changes I've made.

### CODE REVIEW

**Comment by James-Ko** [comment]
Err... looks like something with `ProcessTests.ProcessThreadTests.TestStartTimeProperty` is failing? This is only happening on [Linux debug](http://dotnet-ci.cloudapp.net/job/dotnet_corefx_linux_debug_tst_prtest/1583/).

**Comment by stephentoub** [comment]
> looks like something with ProcessThreadTests.TestStartTimeProperty is failing?

Yeah, https://github.com/dotnet/corefx/issues/2613.  @Priya91, FYI.

File: src/System.Linq/src/System/Linq/Enumerable.cs [hunk 1]
```diff
@@ -703,11 +703,14 @@ public static IEnumerable<TSource> SkipWhile<TSource>(this IEnumerable<TSource>
...
+            using (IEnumerator<TSource> e = source.GetEnumerator())
             {
-                if (!yielding && !predicate(element)) yielding = true;
-                if (yielding) yield return element;
+                do
+                    if (!e.MoveNext()) yield break;
+                while (predicate(e.Current));
+                do
+                    yield return e.Current;
+                while (e.MoveNext());
```
Comment (nguerrera) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: Hmm, the old code was easier to read and maintain IMHO. What kind of speedup do you see here? I'd imagine it wouldn't be very significant vs. general overhead of iterators, but I'm prepared to be proven wrong.

**Comment by stephentoub** [comment]
cc: @VSadov

File: src/System.Linq/src/System/Linq/Enumerable.cs (hunk 1 above)
Comment (Maxwe11) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: +1

Comment (VSadov) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: The new code removes the need for a lifted "yielding" flag an multiple reads of that field. Instead the state machine gets one more state.
This should, in theory, lead to some improvement, but it is indeed curious to know if it indeed made things faster and by how much?

**Comment by stephentoub** [comment]
@dotnet-bot test this please

File: src/System.Linq/src/System/Linq/Enumerable.cs (hunk 1 above)
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: @jamesqo, do you have measurements to go along with this change?

Comment (jamesqo) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: @stephentoub Yep, [here they are](https://gist.github.com/jamesqo/87e3ab99180f4f01cc30) ([results](https://gist.github.com/jamesqo/05aed6873cbd6812b406)). Unfortunately, it looks like nguerrera/Maxwe11 are correct; the new code is barely faster than the old code.

Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: I've found doing similar work that you really have to do these tests across a range of element types, a range of enumerable types and a range of enumerable sizes. The results can vary considerably.

**Comment by stephentoub** [comment]
Thanks, @jamesqo.  Given the discussion thus far and the lack of demonstrable improvements, I'm going to close this out for now.  If you're able to show that this has a real positive impact, please do reopen. Thanks!

**Pull request closed without merging**